pkg/pkgmgr/dpkg.go

```diff
@@ -459,115 +477,96 @@ func (dm *dpkgManager) unpackAndMergeUpdates(ctx context.Context, updates unvers
 			pkgStrings = append(pkgStrings, u.Name)
 		}
-		downloadCmd = fmt.Sprintf(aptGetDownloadTemplate, strings.Join(pkgStrings, " "))
+		downloadCmd = fmt.Sprintf(aptGetDownloadScript, strings.Join(pkgStrings, " "))
 	} else {
-		// only update the outdated pacakges from packages.txt
-		downloadCmd = `
-		packages=$(<packages.txt)
-		for package in $packages; do
-			output=$(apt-get download --no-install-recommends "$package" 2>&1)
-			if [ $? -ne 0 ]; then
-				echo "$output" >>error_log.txt
-			fi
-		done
-		`
+		downloadCmd = aptGetDownloadAllScript
 	}
 
-	downloaded := updated.Dir(dpkgDownloadPath).Run(llb.Args([]string{"bash", "-c", downloadCmd}), llb.WithProxy(utils.GetProxy())).Root()
-
-	// Validate no errors were encountered if updating all
-	if updates == nil && !ignoreErrors {
-		downloaded = downloaded.Run(buildkit.Sh("if [ -s error_log.txt ]; then cat error_log.txt; exit 1; fi")).Root()
+	errorValidation := "false"
+	if ignoreErrors {
+		errorValidation = "true"
 	}
 
-	diffState := llb.Diff(updated, downloaded)
+	// Only need info files and status files for correct installation - copy those.
+	updated = updated.File(llb.Copy(dpkgdb, "/var/lib/dpkg/", "/tmp/debian-rootfs/var/lib/dpkg"))
 
-	// Scripted enumeration and dpkg unpack of all downloaded packages [layer to merge with target]
-	const extractTemplate = `find %s -name '*.deb' -exec dpkg-deb -x '{}' %s \;`
-	extractCmd := fmt.Sprintf(extractTemplate, dpkgDownloadPath, unpackPath)
-	unpacked := downloaded.Run(llb.AddMount(dpkgDownloadPath, diffState), llb.Shlex(extractCmd)).Root()
-	unpackedToRoot := llb.Scratch().File(llb.Copy(unpacked, unpackPath, "/", &llb.CopyInfo{CopyDirContentsOnly: true}))
-
-	// Scripted extraction of all debinfo for version checking to separate layer into local mount
-	// Note that target dirs of shell commands need to be created before use
-	mkFolders := downloaded.File(llb.Mkdir(resultsPath, 0o744, llb.WithParents(true))).File(llb.Mkdir(dpkgStatusFolder, 0o744, llb.WithParents(true)))
-	const writeFieldsTemplate = `find . -name '*.deb' -exec sh -c "dpkg-deb -f {} > %s" \;`
-	writeFieldsCmd := fmt.Sprintf(writeFieldsTemplate, filepath.Join(resultsPath, "{}.fields"))
-	fieldsWritten := mkFolders.Dir(dpkgDownloadPath).Run(llb.Shlex(writeFieldsCmd)).Root()
-
-	// Write the name and version of the packages applied to the results.manifest file for the host
-	const outputResultsTemplate = `find . -name '*.fields' -exec sh -c 'grep "^Package:\|^Version:" {} >> %s' \;`
-	outputResultsCmd := fmt.Sprintf(outputResultsTemplate, resultManifest)
-	resultsWritten := fieldsWritten.Dir(resultsPath).Run(llb.Shlex(outputResultsCmd)).Root()
-	resultsDiff := llb.Diff(fieldsWritten, resultsWritten)
+	// Mount image rootfs into tooling image.
+	// Now, when Copa does dpkg install into the temp rootfs, it wont get override any config files since they are already there.
+	downloaded := updated.Run(
+		llb.AddEnv("IGNORE_ERRORS", errorValidation),
+		buildkit.Sh(downloadCmd),
```

Comment:
I would write the downloadCmd to a file and execute the file which should make the buildkit progress output (on the CLI) easier to grok.
Can use `set -x` in the command if you want to see the full script in the output.

Reply:
Basically anywhere there's a multi-line command being passed into `llb.Run`, the buildkit progress output can be... wonky.